File: 127. Word Ladder/solution-bidirectional-bfs.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int: 
        # edge case
        if beginWord == endWord or not wordList or endWord not in wordList:
            return 0
        
        # create a graph with neighbouring nodes of 1 edit difference
        wordList.append(beginWord)
        graph = {}
        for word in wordList:
            graph[word] = []
        
        for i in range(len(wordList)):
            for j in range(i+1, len(wordList)):
                if self.isOneEdit(wordList[i], wordList[j]):
                    graph[wordList[i]].append(wordList[j])
                    graph[wordList[j]].append(wordList[i])

        # bidirectional BFS
        # maintain visited dictionaries (node -> count)
        v1 = {}
        v2 = {}
        
        # queue for search
        q1 = [beginWord]
        q2 = [endWord]
        
        # count of depth
        c1 = 1
        c2 = 1
        
        # bfs from both sides
        while q1 or q2:
            # bfs on q1
            n1 = []
            for node in q1:
                # should not revisit in this BFS
                if node in v1:
                    continue
                    
                # found overlap
                if node in v2:
                    return c1 + v2[node] - 1
                
                # mark visited
                v1[node] = c1
                
                # add neighbours
                n1.extend(graph[node])
            q1 = n1
            c1 += 1
            
            # bfs on q2
            n2 = []
            for node in q2:
                # should not revisit in this BFS
                if node in v2:
                    continue
                    
                # found overlap
                if node in v1:
                    return c2 + v1[node] - 1
                
                # mark visited
                v2[node] = c2
                
                # add neighbours
                n2.extend(graph[node])
            q2 = n2
            c2 += 1
            
        return 0
            
        
    def isOneEdit(self, word1, word2):
        i = 0
        edits = 0
        
        while i < len(word1):
            if word1[i] != word2[i]:
                edits += 1
                if edits > 1:
                    return False
            i += 1
                
        if edits == 1:
            return True
        else:
            return False
        

File: 127. Word Ladder/solution-bidirectional-bfs.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        wordList.add(beginWord);
        Map<String, Set<String>> transformableWords = getTransformableWords(wordList);
        if (!transformableWords.containsKey(endWord))
            return 0;
        
        return bfs(transformableWords, beginWord, endWord);
    }
    
    public int bfs(Map<String, Set<String>> transformableWords, String beginWord, String endWord) {
        // start bfs from both ends
        int wordCount = transformableWords.size();
        
        // layers containing the current fanout
        Set<String> l1 = new HashSet<String>();
        Set<String> l2 = new HashSet<String>();
        
        // visited set for each end
        Set<String> v1 = new HashSet<String>(wordCount);
        Set<String> v2 = new HashSet<String>(wordCount);
        
        // depth for both ends
        int d1 = 0, d2 = 0;
        
        // init
        l1.add(beginWord);
        // v1.add(beginWord);
        l2.add(endWord);
        // v2.add(endWord);
        
        // run bfs
        while (l1.size()*l2.size() != 0) {
            l1 = getNextLayer(l1, v1, transformableWords);
            d1++;
            
            if (compareLayer(l1, l2)) {
                return d1 + d2 + 1;
            }
            
            l2 = getNextLayer(l2, v2, transformableWords);
            d2++;
            
            if (compareLayer(l1, l2)) {
                return d1 + d2 + 1;
            }            
        }
        
        return 0;
    }
    
    public Set<String> getNextLayer(Set<String> layer, Set<String> visited, Map<String, Set<String>> transformableWords) {
        Set<String> nextLayer = new HashSet<>();
        for (String word: layer) {
            visited.add(word);
            for (String neigh: transformableWords.get(word)) {
                if (!visited.contains(neigh))
                    nextLayer.add(neigh);
            }
        }
        return nextLayer;
    }
    
    public boolean compareLayer(Set<String> l1, Set<String> l2) {
        for (String word: l1) {
            if (l2.contains(word))
                return true;
        }
        return false;
    }
    
    public Map<String, Set<String>> getTransformableWords(List<String> wordList) {
        Map<String, Set<String>> wordToTransformations = new HashMap<String, Set<String>>();
        Set<String> trans;
        for (String word: wordList) {
            trans = new HashSet<String>();
            wordToTransformations.put(word, trans);
        }
        
        String word1, word2;
        for (int i = 0; i < wordList.size(); i++) {
            word1 = wordList.get(i);
            for (int j = i+1; j < wordList.size(); j++) {
                word2 = wordList.get(j);
                if (isOneEditTransformable(word1, word2)) {
                    trans = wordToTransformations.get(word1);
                    trans.add(word2);
                    trans = wordToTransformations.get(word2);
                    trans.add(word1);
                }
            }
        }
        return wordToTransformations;
    }
    
    public boolean isOneEditTransformable(String word1, String word2) {
        int count = 0;
        for (int i = 0; i < word1.length(); i++) {
            if (word1.charAt(i) != word2.charAt(i))
                count++;
            
            if (count > 1) 
                return false;
        }
        
        if (count == 1)
            return true;
        else 
            return false;
    }
}
